1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+    then `TypeResolver` should use the cache associated with ALC-M.
+
+  - Another example, if a `ITypeName` has already resolved the type via reflection in default-ALC,
+    and the `ScriptBlock` containing that `ITypeName` is now being executed in the context of an isolated module M,
+    then the cached type referred by `ITypeName` should be invalided and another resolution should be triggered.
+    Similarly, if the `ScriptBlock` has already been compiled into LINQ expression tree and have the delegate generated,
+    then the cached delegate needs to be voided and another compilation needs to be enforced to make sure type resolution is correct.
+
+For the latter category, theoretically they will continue to work without change,
+because types loaded into different load contexts are different.
+
+### Reclaim Module AssemblyLoadContext
+
+Starting from 3.0, .NET Core supports unloading an `AssemblyLoadContext` and all assemblies in it.
+Ideally, we would like to reclaim the load context associated with an isolated module when the module is removed.
+
+The cache updates called out in the [Update Type Resolution Logic](#update-type-resolution-logic) section
+were designed with the purpose of allowing the custom ALC to be reclaimed.
+By having each custom ALC to hold the cache data structure within itself,
+we won't need to hold a custom ALC or any types loaded in a custom ALC as the key or value for any caches in our engine code,
+and thus there is no need to invalidate any engine cache.
+
+However, we won't be able to guarantee that a custom ALC could be unloaded at all.
+This is because Dynamic Language Runtime (DLR) holds types in its cache, indefinitely.
+
+The DLR types live in the default-ALC,
+and it will generate a `DynamicMetaObject` object for a late-binding operation,
+which is something like a key-value pair `<restrictions, expression>` that will be held in the DLR caches.
+This is a perf improvement feature for dynamic languages like PowerShell because it allows you to skip an expensive late-binding operation when the `restrictions` matches.
+That key-value pair will be held indefinitely,
+so once types from a custom load context is involved in a late binding operation,
+such as property access, method invocation and etc,
+it will be rooted in the GC Heap by the DLR cache, and the custom load context cannot be unloaded afterwards.
+
+## Summary
+
+For the initial implementation, here are the goals and non-goals:
+
+- Goals: work items called out in the [Specification](#specification) section,
+  including updates to the module cmdlets, assembly loading, and type resolution.
+- Non-goals: the new type reference syntax.
+
+To emphasize again:
+The type identity issue is inevitable as soon as different versions of the same assembly can be loaded in the same PowerShell session.
+This will mainly affect modules that expect users to directly create objects of the types loaded by the modules.
```

Comment:
In response to @iSazonov's comment https://github.com/PowerShell/PowerShell-RFC/pull/298#issuecomment-897027194:

The isolated module feature won't solve all problems, and also it won't just make all scenarios automatically work. **The goal, in my opinion, is to offer workaround in a reasonably graceful way for the common scenarios while keeping the complexity under control**. What you said is true, but at the meantime, you need to realize that many modules won't have the type identity issue at all because of how the conflicting assembly gets used.

I already emphasized in the RFC that "the type identity issue is inevitable". But IMHO, it doesn't mean it's un-manageable. It would just require more knowledge about how ALC works and how PowerShell utilizes it. The enhanced type reference syntax would help to mitigate the type identity issue as well.

I think we should be very explicit that a module should be loaded in a separate ALC only if it's needed. We will not encourage blindly using this feature "just for safety". This is important to avoid the type identity issue as much as possible. It's also the main reason why I'm against allowing a module author to control the loading behavior (discussed in [this thread](https://github.com/PowerShell/PowerShell-RFC/pull/298#discussion_r686427648)), because that may be abused and result in modules to be loaded in separate ALC by default.

> I would rather invest in what we used to Windows Compatibility feature - PowerShell remoting serialization. This can be seen as a universal repackaging that gets rid of type conflicts.

It's the same thing for the `-UseWindowsPowerShell` switch -- it won't solve all problems and it won't just make all scenarios automatically work. For example, the object returned will be a deserialized dead object (a property bag) and you won't be able to call methods on it; `cmdletFrom ModuleA | cmdletFrom ModuleB` won't be guaranteed to work either. But it's useful and work sufficiently well for some common scenarios where those unsolved problems are not a concern.

> I suppose we could do it even in-process.

I don't think that's possible, because you will need to load the actual module somewhere, and that cannot be the same process due to the assembly conflicts.

> where modules E1 and E2 have type conflicts. Thus each of the sets could be loaded in separate isolation area. This is very similar to loading modules into different runspaces. This seems much more understandable, predictable (or rather, eliminating unpredictable behavior) and reliable.

Runspace doesn't have assembly level isolation, and you will still get the same assembly loading failure.
If you are talking about having Runspace to be the assembly isolation boundary, then I disagree because it would require working with APIs, which is not friendly to regular interactive shell users.

Reply:
> What you said is true, but at the meantime, 

I wish I had been wrong. 😄 

> you need to realize that many modules won't have the type identity issue at all because of how the conflicting assembly gets used.

This just bewilders me that on rare scenarios we would have to rewrite the entire Engine. I hope we can have more simple way to resolve most of common conflict scenarios.

> It should be the user to decide whether the module should be loaded in a separate load context.

So we can not make PowerShell _smart_ enough to do all work for users and resolve all scenarios. This again raises the question of why we should make so much efforts and add complexity to the Engine and user experience.

-----

If as user I have two conflicting modules (is it main (and maybe single) scenario we want resolve?) ModuleA with assembly 1.0.1 and ModuleB with assembly 1.0.2 I would prefer to be able to **just load** the two modules to make them work (maybe even independently). Is it possible? As we know from the analysis we need to download 1.0.2 first and redirect ModileA to load assembly 1.0.2.

So best user experience for me is to say "PowerShell, when I load ModuleA, please, load assembly 1.0.2. Thanks!".
I'd name this as PowerShell moderated GAC :-)

We can redirect assemblies in .Net applications in its config file. We could do the same for modules too.
If user have install some Az modules and they have dependencies for some Az.Account versions we could ask PowerShell to load highest assembly version.  (If versions are not compatible the modules will do not communicate in ALC isolation too and user have to use separate PowerShell sessions.)
This is much easier to implement than this RFC.

Reply:
No matter how you redirect the loading of an assembly, you still cannot load 2 versions of the same assembly, and thus using a separate ALC is inevitable, and thus all the problems discussed in this RFC remain problems.

Reply:
> It's impossible for PowerShell to know the highest versions of all assemblies. What you are asking is a combination of PowerShell and nuget.org. Even if PowerShell is able to always query and pull from nuget.org, loading the highest version of an assembly is not guaranteed to work with the requesting module because major version number update indicates possible breaking changes.

Oh, no. I don't say PowerShell should be so smart. As you said: "It should be the **user** to decide whether the module should be loaded in a separate load context." - I say too: "It should be the user to know what modules are conflicted"

How the scenario works:
1. User installs two modules (ModuleA and ModuleB) with conflicting assemblies (Assembly1 and Assembly2)
2. User run command like
    - either `Set-ModuleOptions -Modules @("ModuleA", "ModuleB") -Assemblies @("Assembly1", "Assembly2")`
    - or `Set-ModuleOptions -Modules @("ModuleA", "ModuleB")` - detect common assemblies from module manifests
Notice, if user updates one of the modules the command in user profile automatically updates the config.
3. PowerShell analyzes the modules and finds highest version of common assemblies
4. User run `Import-Module ModuleA` or `Import-Module ModuleB` and PowerShell loads right assembly.

Reply:
> detect common assemblies from module manifests

Assembly loading could happen from a .psm1, and that's quite common.

> Notice, if user updates one of the modules **the command in user profile** automatically updates the config.

Having this command in user's profile? Asking the user to always know what modules they will use for every session?

I think this user experience is very unintuitive.

Reply:
This seems too contradictory. If we say it's the user's responsibility to properly use conflict resolution in a limited set of scenarios, we can't say "unintuitive", especially since the features in this RFC are certainly not intuitive.

And if we're trying to alleviate the conflict problem, why don't the authors of the modules make an effort as well? Either avoid problematic dependencies, or hide them, or declare assemblies in manifest to allow users to use it more easily and correctly. 
And this is not necessarily a requirement in my proposal: the user can explicitly specify the assemblies and modules that require special handling.
------

A key in my proposal whether loading an assembly with high version is sufficient to address the necessary scenarios?

Reply:
> especially since the features in this RFC are certainly not intuitive.

This RFC proposes 2 new parameters to `Import-Module`. When a user ran into assembly conflicting issue when loading a module, they just need to decide whether they want to load that module into a different ALC. That's much more intuitive than what you are suggesting, especially it may not obvious for a user to know which modules introduce this assembly incompatibility.

> And if we're trying to alleviate the conflict problem, why don't the authors of the modules make an effort as well? Either avoid problematic dependencies, or hide them, or declare assemblies in manifest to allow users to use it more easily and correctly.

I already mentioned that would be an area for PowerShellGet and PowerShell Gallery vNext to look into. And I quoted it here again:
> However, we need to clearly realize that there won't be a single solution that can solve all the problems and make all scenarios work automatically. We actually discussed the same idea internally before, and this is an area to look more into for the PowerShellGet and PowerShell Gallery vNext. This idea doesn't go against the isolated module approach, and it actually helps to prevent from abusing the isolated module approach so as to reduce the risk of the type identity issue.

Reply:
> This RFC proposes 2 new parameters to Import-Module. When a user ran into assembly conflicting issue when loading a module, they just need to decide whether they want to load that module into a different ALC, with one of the 2 new parameters. That's much more intuitive than what you are suggesting, especially it may not be obvious in many cases for a user to know which modules introduce this assembly incompatibility.

I can not agree with such "intuitive". "Intuitive" for me if PowerShell could load and use modules without conflicts at all but we can not achieve it.

Let's start with original issue. What is a message users get?
```
Import-Module: Assembly with same name is already loaded
```
Can we make the message more useful? Yes, we can.
Do we know a module we try to load? Yes, and we can include the module name in the message.
Do we know the assembly name and version from the module? Yes, and we can include the name in the message.
Do we know module(s) and assemblies which was loaded early and cause the conflict? We can find them and include in the message.

Thus, we can write great suggestion message to user how he/she could to try resolve the conflict like "Try load <D1> module(s) before module <B1>" or "Try load <A1> assembly with 'Add-Type -Assembly A1' before loading modules <D1> and <B1>".

I think loading assembly with highest version resolves all common scenarios in better way and more scenarios than with ALCs.
I expect this will work well if there is only difference in build and minor assembly version numbers. If there are incompatibilities in APIs they can not be resolved with ALCs too. Moreover, with ALCs there will be more opportunities for problems to arise. 
Only way to try to resolve these API incompatibilities is to use WindowsCompatibility approach to run another process with serialization.

Then I'd can speculate if the RFC would come from community PowerShell Committee conclusion would be "We have full consensus and agree to improve the message but we reject the RFC because:
- it does not resolve all scenarios
- we are very small team and haven't resources for so large work
- this implies an unacceptably high risk of bugs because almost all subsystems of the engine will be affected"

I am sure that the implementation of this RFC is really not worth the effort out of high cost, high risk and only a partial solution to the problem.

I would start by improving said message. This is a one hour job for MSFT team.
Then I would agree that loading an assembly with highest version is the best solution - simple, fast, reliable and address more scenarios.

Then, we have some options to go.
- Since MSFT team always say "we don't want to increase Engine size,  please create community module first." we could create a standalone analyzer/module to help users to find conflicting modules.
- We could makes PowerShell Engine itself more smart, add new cmdlet for the analyzing, add the info in module cache and so on.

Now after I've refreshed the issue history and I think we should start with improving the error message and wait feedback. Probably nothing else will be needed.
If the activity with this RFC is related to AZ module problems MSFT could ensure annotating used assemblies in module manifests and we could easily learn PowerShell to analyze this and load assemblies with highest versions.

Reply:
> Do we know module(s) and assemblies which was loaded early and cause the conflict? We can find them and include in the message.

I'm afraid you cannot find out this information unless untrivial amount of work is done in engine to bookkeep this information along the way.

> Thus, we can write great suggestion message to user how he/she could to try resolve the conflict like "Try load module(s) before module " or "Try load assembly with 'Add-Type -Assembly A1' before loading modules and ".

Then what, asking the user to exit the current session and restart? They cannot reordering the loading in the current session as it's too late.

I respect your opinions but I don't agree with your opinion on this.

Reply:
> I'm afraid you cannot find out this information unless untrivial amount of work is done in engine to bookkeep this information along the way.

We can catch the exception in module loading code where we know the module we load.

> Then what, asking the user to exit the current session and restart? They cannot reorder the loading in the current session as it's too late.

Yes, of cause. It is an investigation time when user try to understand problem and find best for him solution.
This process is exactly the same as for any other error.
If we want more convenience and to load the correct assembly at once, we could store this information with the module  cache and update it (for example, when a new module is added to the cache)